File: app/services/workflow.py
import logging
from datetime import datetime
from typing import Any

from app.connector.factory import ConnectorFactory
from app.schemas.workflow import StepResult, WorkflowDefinition, WorkflowStep
from app.schemas.run import WorkflowRun
from app.schemas.common import StepStatus, WorkflowStatus
from app.repositories.workflow import WorkflowRepository
from app.repositories.run import WorkflowRunRepository
from app.storage.enum import StorageType
from app.storage.factory import StorageFactory


logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class WorkflowService:
    def __init__(self, storage: StorageType):
        """
        Initialize the WorkflowService with a specific storage type.

        Args:
            storage (StorageType): The type of storage to use (e.g., StorageType.FILE_SYSTEM).

        Raises:
            ValueError: If the storage type is unknown.
        """
        storage_cls = StorageFactory.create_storage(storage)

        # Initialize storage and engine
        workflow_storage = storage_cls[WorkflowDefinition](t_type=WorkflowDefinition)
        workflow_run_storage = storage_cls[WorkflowRun](t_type=WorkflowRun)

        self.workflow_repository = WorkflowRepository(workflow_storage)
        self.workflow_run_repository = WorkflowRunRepository(workflow_run_storage)

    def create_workflow(self, workflow: WorkflowDefinition) -> str:
        """
        Create a new workflow definition.

        Args:
            workflow (WorkflowDefinition): The workflow definition to create.

        Returns:
            str: The UUID of the created workflow.
        """
        return self.workflow_repository.create_workflow(workflow)

    def load_workflow(self, uuid: str) -> WorkflowDefinition:
        """
        Load a workflow definition by its UUID.

        Args:
            uuid (str): The UUID of the workflow to load.

        Returns:
            WorkflowDefinition: The workflow definition, or None if not found.
        """
        return self.workflow_repository.get_workflow(uuid)

    def create_workflow_run(self, workflow_run: WorkflowRun) -> str:
        """
        Create a new workflow run.

        Args:
            workflow_run (WorkflowRun): The workflow run object to save.

        Returns:
            str: The UUID of the created workflow run.
        """
        return self.workflow_run_repository.create_workflow_run(workflow_run)

    def load_workflow_run(self, uuid: str) -> WorkflowRun:
        """
        Load a workflow run by its UUID.

        Args:
            uuid (str): The UUID of the workflow run to load.

        Returns:
            WorkflowRun: The workflow run object, or None if not found.
        """
        return self.workflow_run_repository.get_workflow_run(uuid)

    def list_runs(self) -> list[WorkflowRun]:
        """
        List all workflow runs.

        Returns:
            list[WorkflowRun]: A list of all workflow runs.
        """
        return self.workflow_run_repository.list_workflow_runs()

    async def execute_workflow(self, run_id: str):
        """
        Execute a workflow run.

        Args:
            run_id (str): The UUID of the workflow run to execute.

        This method:
            1. Loads the run and workflow definition.
            2. Updates status to RUNNING.
            3. Iterates through steps sequentially, passing context.
            4. Handling step retries (if implemented) or failure.
            5. Updates final status to SUCCESS or FAILED.
        """
        # This method would contain the logic to execute the workflow.
        # For now, we will just return a placeholder message.
        run = self.load_workflow_run(run_id)
        if not run:
            logger.error(f"Workflow run {run_id} not found")
            return

        workflow = self.load_workflow(run.workflow_id)
        if not workflow:
            logger.error(f"Workflow {run.workflow_id} not found")
            run.status = WorkflowStatus.FAILED
            run.error = f"Workflow {run.workflow_id} not found"
            self.workflow_run_repository.update_workflow_run(run)
            return

        logger.info(f"Executing workflow run {run_id}")
        run.status = WorkflowStatus.RUNNING
        self.workflow_run_repository.update_workflow_run(run)

        context = {"payload": run.payload}
        try:
            for step in workflow.steps:
                step_result = await self._execute_step(step, context)
                run.step_results[step.name] = step_result

                if step_result.status == StepStatus.FAILED:
                    run.status = WorkflowStatus.FAILED
                    run.error = step_result.error
                    run.completed_at = datetime.now().isoformat()
                    self.workflow_run_repository.update_workflow_run(run)
                    return

                # Add step output to context for next steps
                if step_result.output:
                    context[step.name] = step_result.output

            # All steps completed successfully
            run.status = WorkflowStatus.SUCCESS
            run.completed_at = datetime.now().isoformat()
            self.workflow_run_repository.update_workflow_run(run)
            logger.info(f"Workflow run {run_id} completed successfully")

        except Exception as e:
            logger.error(f"Workflow run {run_id} failed: {str(e)}")
            run.status = WorkflowStatus.FAILED
            run.error = str(e)
            run.completed_at = datetime.now().isoformat()
            self.workflow_run_repository.update_workflow_run(run)

    async def _execute_step(self, step: WorkflowStep, context: dict[str, Any]):
        """
        Execute a single workflow step.

        Args:
            step (WorkflowStep): The step definition.
            context (dict[str, Any]): The execution context containing payload and previous step outputs.

        Returns:
            StepResult: The result of the step execution.
        """
        """This method would contain the logic to execute a single step."""
        result = StepResult(
            step_name=step.name,
            status=StepStatus.RUNNING,
            started_at=datetime.now().isoformat(),
        )
        try:
            # Simulate step execution
            connector = ConnectorFactory.get_instance(step.type)
            logger.info(f"Executing step: {step.name} ({step.type})")
            result.output = await connector.execute(step, context)
            result.status = StepStatus.SUCCESS
            result.completed_at = datetime.now().isoformat()
        except Exception as e:
            logger.error(f"Step {step.name} failed: {str(e)}")
            result.status = StepStatus.FAILED
            result.error = str(e)
            result.completed_at = datetime.now().isoformat()
        return result

File: rust-app/src/services/workflow.rs
use std::collections::HashMap;
use std::sync::Arc;
use tracing::{error, info};

use crate::connector::factory::ConnectorFactory;
use crate::repositories::run::WorkflowRunRepository;
use crate::repositories::workflow::WorkflowRepository;
use crate::schemas::common::{StepStatus, WorkflowStatus};
use crate::schemas::run::WorkflowRun;
use crate::schemas::workflow::{StepResult, WorkflowDefinition};
use crate::storage::base::Storage;

/// Service layer for workflow operations.
///
/// Mirrors Python's WorkflowService from app/services/workflow.py.
/// Follows Single Responsibility: orchestrates workflow CRUD and execution
/// by delegating to repositories and connectors.
/// Follows Dependency Inversion: depends on Storage trait abstractions, not concrete types.
pub struct WorkflowService {
    workflow_repository: WorkflowRepository,
    workflow_run_repository: WorkflowRunRepository,
}

impl WorkflowService {
    /// Initialize the WorkflowService with storage backends.
    ///
    /// # Arguments
    /// * `workflow_storage` - Storage backend for workflow definitions.
    /// * `run_storage` - Storage backend for workflow runs.
    pub fn new(
        workflow_storage: Arc<dyn Storage<WorkflowDefinition>>,
        run_storage: Arc<dyn Storage<WorkflowRun>>,
    ) -> Self {
        Self {
            workflow_repository: WorkflowRepository::new(workflow_storage),
            workflow_run_repository: WorkflowRunRepository::new(run_storage),
        }
    }

    /// Create a new workflow definition.
    pub async fn create_workflow(
        &self,
        workflow: &mut WorkflowDefinition,
    ) -> Result<String, crate::storage::base::StorageError> {
        self.workflow_repository.create_workflow(workflow).await
    }

    /// Load a workflow definition by its UUID.
    pub async fn load_workflow(
        &self,
        uuid: &str,
    ) -> Result<Option<WorkflowDefinition>, crate::storage::base::StorageError> {
        self.workflow_repository.get_workflow(uuid).await
    }

    /// Create a new workflow run.
    pub async fn create_workflow_run(
        &self,
        run: &mut WorkflowRun,
    ) -> Result<String, crate::storage::base::StorageError> {
        self.workflow_run_repository.create_workflow_run(run).await
    }

    /// Load a workflow run by its UUID.
    pub async fn load_workflow_run(
        &self,
        uuid: &str,
    ) -> Result<Option<WorkflowRun>, crate::storage::base::StorageError> {
        self.workflow_run_repository.get_workflow_run(uuid).await
    }

    /// List all workflow runs.
    pub async fn list_runs(
        &self,
    ) -> Result<Vec<WorkflowRun>, crate::storage::base::StorageError> {
        self.workflow_run_repository.list_workflow_runs().await
    }

    /// List workflow runs with cursor-based pagination.
    pub async fn list_runs_paginated(
        &self,
        limit: i64,
        cursor: Option<&str>,
    ) -> Result<(Vec<WorkflowRun>, Option<String>), crate::storage::base::StorageError> {
        self.workflow_run_repository
            .list_workflow_runs_paginated(limit, cursor)
            .await
    }

    /// Update a workflow run.
    pub async fn update_workflow_run(
        &self,
        run: &WorkflowRun,
    ) -> Result<bool, crate::storage::base::StorageError> {
        self.workflow_run_repository.update_workflow_run(run).await
    }

    /// Execute a workflow run.
    ///
    /// This method:
    /// 1. Loads the run and workflow definition.
    /// 2. Updates status to RUNNING.
    /// 3. Iterates through steps sequentially, passing context.
    /// 4. Updates final status to SUCCESS or FAILED.
    pub async fn execute_workflow(&self, run_id: &str) {
        let run = match self.load_workflow_run(run_id).await {
            Ok(Some(r)) => r,
            Ok(None) => {
                error!("Workflow run {} not found", run_id);
                return;
            }
            Err(e) => {
                error!("Error loading workflow run {}: {}", run_id, e);
                return;
            }
        };

        let workflow = match self.load_workflow(&run.workflow_id).await {
            Ok(Some(w)) => w,
            Ok(None) => {
                error!("Workflow {} not found", run.workflow_id);
                let mut failed_run = run;
                failed_run.status = WorkflowStatus::Failed;
                failed_run.error = Some(format!("Workflow {} not found", failed_run.workflow_id));
                let _ = self.update_workflow_run(&failed_run).await;
                return;
            }
            Err(e) => {
                error!("Error loading workflow {}: {}", run.workflow_id, e);
                return;
            }
        };

        info!("Executing workflow run {}", run_id);
        let mut run = run;
        run.status = WorkflowStatus::Running;
        let _ = self.update_workflow_run(&run).await;

        let mut context: HashMap<String, serde_json::Value> = HashMap::new();
        context.insert(
            "payload".to_string(),
            serde_json::to_value(&run.payload).unwrap_or_default(),
        );

        for step in &workflow.steps {
            let step_result = self.execute_step(step, &context).await;

            if step_result.status == StepStatus::Failed {
                run.status = WorkflowStatus::Failed;
                run.error = step_result.error.clone();
                run.completed_at = Some(chrono::Utc::now().to_rfc3339());
                run.step_results
                    .insert(step.name().to_string(), step_result);
                let _ = self.update_workflow_run(&run).await;
                return;
            }

            // Add step output to context for next steps
            if let Some(ref output) = step_result.output {
                if let Ok(val) = serde_json::to_value(output) {
                    context.insert(step.name().to_string(), val);
                }
            }

            run.step_results
                .insert(step.name().to_string(), step_result);
        }

        // All steps completed successfully
        run.status = WorkflowStatus::Success;
        run.completed_at = Some(chrono::Utc::now().to_rfc3339());
        let _ = self.update_workflow_run(&run).await;
        info!("Workflow run {} completed successfully", run_id);
    }

    /// Execute a single workflow step.
    async fn execute_step(
        &self,
        step: &crate::schemas::workflow::WorkflowStep,
        context: &HashMap<String, serde_json::Value>,
    ) -> StepResult {
        let mut result = StepResult {
            step_name: step.name().to_string(),
            status: StepStatus::Running,
            started_at: chrono::Utc::now().to_rfc3339(),
            completed_at: None,
            output: None,
            error: None,
        };

        let connector_type = step.connector_type();
        match ConnectorFactory::get_instance(&connector_type) {
            Ok(connector) => {
                info!("Executing step: {} ({:?})", step.name(), connector_type);
                match connector.execute(step, context).await {
                    Ok(output) => {
                        result.output = Some(output);
                        result.status = StepStatus::Success;
                        result.completed_at = Some(chrono::Utc::now().to_rfc3339());
                    }
                    Err(e) => {
                        error!("Step {} failed: {}", step.name(), e);
                        result.status = StepStatus::Failed;
                        result.error = Some(e.to_string());
                        result.completed_at = Some(chrono::Utc::now().to_rfc3339());
                    }
                }
            }
            Err(e) => {
                error!("Failed to get connector for step {}: {}", step.name(), e);
                result.status = StepStatus::Failed;
                result.error = Some(e);
                result.completed_at = Some(chrono::Utc::now().to_rfc3339());
            }
        }

        result
    }
}
